File: solutions/0008-string-to-integer-atoi/solution.java
class Solution {
    public int myAtoi(String s) {
        int sign = 1;
        
        String t = s.trim();
        String u;
        
        if(t.length() == 0) return 0;
        else if(t.charAt(0) == '+') {
            sign = 1;
            u = t.substring(1);
        }    
        else if(t.charAt(0) == '-') {
            sign = -1;
            u = t.substring(1);
        }    
        else if(!Character.isDigit(t.charAt(0))) {
            return 0;
        }  
        else u = t;
        
        int result = 0;
        int i = 0;
        int n = u.length();
        
        while(i < n && Character.isDigit(u.charAt(i))) {
            int pop = Character.getNumericValue(u.charAt(i));

            if(sign == 1) {
                if(result == Integer.MAX_VALUE/10 && pop > 7) {
                    return Integer.MAX_VALUE;
                }
                
                else if(result > Integer.MAX_VALUE/10) {
                    return Integer.MAX_VALUE;
                }
            }

            if(sign == -1) {
                if(-1*result == Integer.MIN_VALUE/10 && pop > 8) {
                    return Integer.MIN_VALUE;
                }
                
                else if(-1*result < Integer.MIN_VALUE/10) {
                    return Integer.MIN_VALUE;
                }
            }

            result = result * 10;
            result = result + pop;
            i++;
        }  
        
        return (sign*result);
        

    }
}

File: solutions/0008-string-to-integer-atoi/solution.js
/**
 * @param {string} s
 * @return {number}
 */
var myAtoi = function (s) {
    s = s.trim();  // Remove leading/trailing whitespace
    let sign = 1;   // Default sign is positive
    let result = 0; // This will store the resulting integer

    let i = 0;

    // Check for sign
    if (s[i] === '-' || s[i] === '+') {
        sign = (s[i] === '-') ? -1 : 1;
        i++;  // Move past the sign
    }

    // Convert digits to number
    while (i < s.length && !isNaN(s[i]) && s[i] >= '0' && s[i] <= '9') {
        result = result * 10 + (s[i] - '0');
        i++;
    }

    // Apply the sign
    result *= sign;

    // Handle overflow and underflow
    const INT_MAX = 2147483647;
    const INT_MIN = -2147483648;

    if (result > INT_MAX) return INT_MAX;
    if (result < INT_MIN) return INT_MIN;

    return result;
};

